### PR TITLE
[breaking-change] add safeguards to public methods

### PR DESCRIPTION
**Pull request by Lotphy**
### Description

Add safeguards to vast-tracker public methods according to IAB doc.

### Issue

Public methods of vast tracker are lacking security for input params. We need to make sure that the inputs type meets the IAB standards for every public method that we expose in the vast tracker.

### Type
- [x] Breaking change
- [ ] Enhancement
- [ ] Fix
- [ ] Documentation
- [ ] Tooling

### CODE REVIEW

File: src/vast_tracker.js [hunk 1]
```diff
@@ -225,6 +236,13 @@ export class VASTTracker extends EventEmitter {
    * @return {Boolean}
    */
   isQuartileReached(quartile, time, progress) {
```
Comment (ZacharieTFR) [inline src/vast_tracker.js:238]: Not sure if it's usefull to add a safeguard here, as `isQuartileReached` is not part of [public methods](https://github.com/dailymotion/vast-client-js/blob/master/docs/api/vast-tracker.md#public-methods--) and its inputs are already covered in `setProgress` method

File: src/vast_tracker.js [hunk 2]
```diff
@@ -173,14 +180,18 @@ export class VASTTracker extends EventEmitter {
...
     const skipDelay = this.skipDelay || DEFAULT_SKIP_DELAY;
 
     if (skipDelay !== -1 && !this.skippable) {
       if (skipDelay > progress) {
-        this.emit('skip-countdown', skipDelay - progress);
+        this.emit('skip-countdown', [skipDelay - progress]);
       } else {
         this.skippable = true;
-        this.emit('skip-countdown', 0);
+        this.emit('skip-countdown', [0]);
```
Comment (ZacharieTFR) [inline src/vast_tracker.js:194]: Nice 👍

**Review by ZacharieTFR — CHANGES_REQUESTED** [review]

File: src/vast_tracker.js [hunk 3]
```diff
@@ -571,6 +641,9 @@ export class VASTTracker extends EventEmitter {
    *
    */
   track(eventName, { macros = {}, once = false } = {}) {
+    if (typeof eventName !== 'string' || typeof macros !== 'object') {
```
Comment (ZacharieTFR) [inline src/vast_tracker.js:644]: Not sure if the test is usefull as we already do some checks on `eventName` below

File: src/util/util.js [hunk 4]
```diff
@@ -32,7 +32,7 @@ function resolveURLTemplates(URLTemplates, macros = {}, options = {}) {
   }
 
   // Calc random/time based macros
-  macros['CACHEBUSTING'] = leftpad(
+  macros['CACHEBUSTING'] = addLeadingZeros(
```
Comment (GuillaumeBAECHLER) [inline src/util/util.js:35]: Won't `Math.round(Math.random() * 1.0e8).toString().padStart(8, '0')` do the job here ? 🤔

File: src/vast_tracker.js [hunk 5]
```diff
@@ -536,6 +603,9 @@ export class VASTTracker extends EventEmitter {
    * @emits VASTTracker#clickthrough
    */
   click(fallbackClickThroughURL = null, macros = {}) {
+    if (fallbackClickThroughURL !== 'string' || typeof macros !== 'object') {
```
Comment (GuillaumeBAECHLER) [inline src/vast_tracker.js:606]: Missing `typeof` ?

File: src/vast_tracker.js [hunk 6]
```diff
@@ -3,6 +3,7 @@ import { isCreativeLinear } from './creative/creative_linear';
 import { EventEmitter } from './util/event_emitter';
 import { isNonLinearAd } from './non_linear_ad';
 import { util } from './util/util';
+import {type} from 'mocha/lib/utils.js';
```
Comment (GuillaumeBAECHLER) [inline src/vast_tracker.js:6]: auto import ? 🙈

**Comment by GuillaumeBAECHLER** [comment]
Nice! A must have for sure :muscle:
Just wondering if we shouldn't add some kind of log in case we have a wrong input.
Maybe this can be helpful for debuging if some trackers are not working ? 🤔

File: src/util/util.js (hunk 4 above)
Comment (Lotphy) [reply src/util/util.js:35]: I didn't know such a thing existed :o

File: src/util/util.js [hunk 7]
```diff
@@ -32,9 +32,7 @@ function resolveURLTemplates(URLTemplates, macros = {}, options = {}) {
   }
 
   // Calc random/time based macros
-  macros['CACHEBUSTING'] = leftpad(
-    Math.round(Math.random() * 1.0e8).toString()
-  );
+  macros['CACHEBUSTING'] = Math.round(Math.random() * 1.0e8).toString().padStart(8, '0');
```
Comment (ZacharieTFR) [inline src/util/util.js:35]: You should call `addLeadingZeros` here 😉

**Review by ZacharieTFR — APPROVED** [review]

File: src/vast_tracker.js [hunk 8]
```diff
@@ -305,9 +325,10 @@ export class VASTTracker extends EventEmitter {
    * @param {Number} duration - The time in seconds until the skip button is displayed.
    */
   setSkipDelay(duration) {
-    if (typeof duration === 'number') {
-      this.skipDelay = duration;
+    if (!util.isValidTimeValue(duration) || typeof macros !== 'object') {
```
Comment (GuillaumeBAECHLER) [inline src/vast_tracker.js:328]: I think checking macros here will break the `setSkipDelay` function 😅

**Review by GuillaumeBAECHLER — CHANGES_REQUESTED** [review]

**Review by GuillaumeBAECHLER — APPROVED** [review]
LGTM 🚢

**Review by ZacharieTFR — APPROVED** [review]

**Pull request merged**